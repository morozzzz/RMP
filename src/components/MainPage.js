import React from 'react';
import TopMenu from './TopMenu';
import MovieList from './MovieList';
// import Footer from './Footer';


class MainPage extends React.Component {
    constructor() {
        super();
        this.state = {
            movies: [
                {
                    title: 'sdfsdf sdfsdffsdf',
                    imageURL: 'https://m.media-amazon.com/images/M/MV5BMTQ2NzYxNjQyMF5BMl5BanBnXkFtZTcwOTQ0OTI0Mw@@._V1_.jpg',
                    genre: 'sfs sdfsdf',
                    year: '1555',
                    id: 1,
                },
                {
                    title: 'sdfsdf sdfsdffsdf',
                    imageURL: 'https://m.media-amazon.com/images/M/MV5BMTQ2NzYxNjQyMF5BMl5BanBnXkFtZTcwOTQ0OTI0Mw@@._V1_.jpg',
                    genre: 'sfs sdfsdf',
                    year: '1555',
                    id: 2,
                },
                {
                    title: 'sdfsdf sdfsdffsdf',
                    imageURL: 'https://m.media-amazon.com/images/M/MV5BMTQ2NzYxNjQyMF5BMl5BanBnXkFtZTcwOTQ0OTI0Mw@@._V1_.jpg',
                    genre: 'sfs sdfsdf',
                    year: '1555',
                    id: 3,
                },
                {
                    title: 'sdfsdf sdfsdffsdf',
                    imageURL: 'https://m.media-amazon.com/images/M/MV5BMTQ2NzYxNjQyMF5BMl5BanBnXkFtZTcwOTQ0OTI0Mw@@._V1_.jpg',
                    genre: 'sfs sdfsdf',
                    year: '1555',
                    id: 4,
                },
            ],
        };
    }

    render() {
        const { movies } = this.state;
        return (
            <div className="main-page">
                <TopMenu />
                <MovieList movies={movies} />
            </div>
        );
    }
}

export default MainPage;
