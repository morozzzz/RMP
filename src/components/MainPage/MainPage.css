.main-page {
    min-height: 100%;
    display: grid;
    grid-template-rows: auto auto 1fr auto;
    grid-template-columns: 100%;
}