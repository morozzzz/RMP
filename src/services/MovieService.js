import { GET_MOVIES_URL } from '../constants/app.constants';


const getMovies = (options = {}) => {
    const queryString = Object.keys(options)
        .map(param => `${param}=${options[param]}&`)
        .join('')
        .slice(0, -1);

    const targetUrl = `${GET_MOVIES_URL}?${queryString}`;

    return fetch(targetUrl);
};

const getMovieById = () => Promise.resolve({}); // todo

export {
    getMovies,
    getMovieById,
};
