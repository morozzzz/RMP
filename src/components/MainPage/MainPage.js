import React from 'react';
import TopMenu from '../TopMenu/TopMenu';
import MovieList from '../MovieList/MovieList';
import StatusBar from '../StatusBar/StatusBar';
import SortPanel from '../SortPanel/SortPanel';
import Footer from '../Footer/Footer';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';
import './MainPage.css';

class MainPage extends React.Component {
    constructor() {
        super();
        this.state = {
            movies: [],
        };
    }

    onPosterClick = () => {
        console.log(123);
    }

    render() {
        const { movies } = this.state;
        return (
            <ErrorBoundary>
                <div className="main-page">
                    <TopMenu />
                    <StatusBar status={movies.length ? `${movies.length} movies found` : ''}>
                        {
                            movies.length ? <SortPanel /> : null
                        }
                    </StatusBar>
                    <MovieList movies={movies} onItemClick={this.onPosterClick} />
                    <Footer />
                </div>
            </ErrorBoundary>
        );
    }
}

export default MainPage;
