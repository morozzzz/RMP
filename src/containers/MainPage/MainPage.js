import React from 'react';
import { connect } from 'react-redux';
import queryString from 'query-string';
import TopMenu from '../../components/TopMenu/TopMenu';
import MovieList from '../../components/MovieList/MovieList';
import StatusBar from '../../components/StatusBar/StatusBar';
import SortPanel from '../SortPanel/SortPanel';
import Footer from '../../components/Footer/Footer';
import ErrorBoundary from '../../components/ErrorBoundary/ErrorBoundary';
import {
    fetchMoviesByCriteria,
} from '../../actions/movies.actions';
import './MainPage.css';

class MainPage extends React.Component {
    constructor(props) {
        super();
        this.getMovies = props.getMovies.bind(this);
    }

    componentDidMount = () => {
        const { location } = this.props;
        const { search } = location;

        if (!search) return;

        this.getMovies(search);
    }

    onPosterClick = (data) => {
        const { history } = this.props;

        history.push(`/film/${data.id}`);
    }

    updateMovies = () => {
        const { searchParams, history } = this.props;

        const query = queryString.stringify(searchParams);

        history.push(`/search/?${query}`);

        this.getMovies(query);
    }

    render() {
        const { movies, isMoviesFound } = this.props;

        return (
            <ErrorBoundary>
                <div className="main-page">
                    <TopMenu onSearchClick={this.updateMovies} />
                    <StatusBar status={isMoviesFound ? `${movies.length} movies found` : ''}>
                        {
                            isMoviesFound ? <SortPanel onSortUpdate={this.updateMovies} /> : null
                        }
                    </StatusBar>
                    {
                        typeof isMoviesFound === 'boolean' ? <MovieList movies={movies} onItemClick={this.onPosterClick} /> : <div />
                    }
                    <Footer />
                </div>
            </ErrorBoundary>
        );
    }
}

const mapStateToProps = state => (
    {
        movies: state.movies.general,
        searchParams: state.searchParams,
        isMoviesFound: state.movies.isMoviesFound,
    }
);

const mapDispatchToProps = dispatch => (
    {
        getMovies: options => dispatch(fetchMoviesByCriteria(options)),
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(MainPage);
