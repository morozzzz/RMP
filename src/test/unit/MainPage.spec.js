import React from 'react';
import { shallow } from 'enzyme';
import MainPage from '../../components/MainPage/MainPage';
import MovieService from '../../services/MovieService';
import '@babel/polyfill';

const moviesMock = [
    {
        id: 333339,
        title: 'Ready Player One',
        tagline: 'A better reality awaits.',
        vote_average: 8.1,
        vote_count: 617,
        release_date: '2018-03-28',
        poster_path: 'https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg',
        overview: 'When the creator of a popular video game system dies, a virtual contest is created to compete for his fortune.',
        budget: 175000000,
        revenue: 0,
        genres: [
            'Adventure',
            'Science Fiction',
            'Action',
        ],
        runtime: 140,
    },
    {
        id: 338970,
        title: 'Tomb Raider',
        tagline: 'Her legend begins',
        vote_average: 6.2,
        vote_count: 817,
        release_date: '2018-03-08',
        poster_path: 'https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg',
        overview: 'Lara Croft, the fiercely independent daughter of a missing adventurer, must push herself beyond her limits when she finds herself on the island where her father disappeared.',
        budget: 94000000,
        revenue: 126025000,
        genres: [
            'Action',
            'Adventure',
        ],
        runtime: 118,
    },
];

jest.mock('../../services/MovieService');
jest.mock('../../components/TopMenu/TopMenu', () => () => <div id="TopMenu" />);
jest.mock('../../components/MovieList/MovieList', () => () => <div id="MovieList" />);
jest.mock('../../components/StatusBar/StatusBar', () => () => <div id="StatusBar" />);
jest.mock('../../components/SortPanel/SortPanel', () => () => <div id="SortPanel" />);
jest.mock('../../components/Footer/Footer', () => () => <div id="Footer" />);
jest.mock('../../components/ErrorBoundary/ErrorBoundary', () => () => <div id="ErrorBoundary" />);

describe.skip('MainPage', () => {
    beforeEach(() => {
        MovieService.getMovies.mockImplementation(() => Promise.resolve(moviesMock));
    });

    it('should be rendered correctly', () => {
        const wrapper = shallow(<MainPage />);

        expect(wrapper).toMatchSnapshot();
    });

    it('componentDidMount() should set state with movies', () => {
        const wrapper = shallow(<MainPage />);
        const instance = wrapper.instance();

        instance.componentDidMount()
            .then(() => {
                expect(wrapper.state().movies).toEqual(moviesMock);
            });
    });

    it('updateMovies() should set state with new movies', () => {
        const newMovies = [
            {
                id: 181808,
                title: 'Star Wars: The Last Jedi',
                tagline: 'The Saga Continues',
                vote_average: 7.1,
                vote_count: 4732,
                release_date: '2017-12-13',
                poster_path: 'https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg',
                overview: 'Rey develops her newly discovered abilities with the guidance of Luke Skywalker, who is unsettled by the strength of her powers. Meanwhile, the Resistance prepares to do battle with the First Order.',
                budget: 200000000,
                revenue: 1325937250,
                genres: [
                    'Fantasy',
                    'Adventure',
                    'Science Fiction',
                ],
                runtime: 152,
            },
        ];

        MovieService.getMovies.mockImplementation(() => Promise.resolve(newMovies));

        const wrapper = shallow(<MainPage />);
        const instance = wrapper.instance();

        instance.updateMovies()
            .then(() => {
                expect(wrapper.state().movies).toEqual(newMovies);
            });
    });
});
