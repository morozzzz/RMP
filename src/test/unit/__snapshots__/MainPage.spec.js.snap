// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should be rendered correctly 1`] = `
<Component>
  <div
    className="main-page"
  >
    <Component
      onSearchClick={[Function]}
    />
    <Component
      status=""
    />
    <Component
      movies={Array []}
      onItemClick={[Function]}
    />
    <Component />
  </div>
</Component>
`;
